### PR TITLE
don't lint on x = x + y inside an AddAssign impl

### LINKED ISSUE #1302: assign_op_pattern false positive for Assign trait impl
We start out with

``` rust
use std::ops::{Mul, MulAssign};

#[derive(Copy, Clone, Debug, PartialEq)]
pub struct Wrap(i64);

impl Mul<i64> for Wrap {
    type Output = Self;

    fn mul(self, rhs: i64) -> Self {
        Wrap(self.0 * rhs)
    }
}

impl MulAssign<i64> for Wrap {
    fn mul_assign(&mut self, rhs: i64) {
        *self = *self * rhs
    }
}

#[test]
fn test() {
    let mut foo = Wrap(5);
    foo *= 10;
    assert_eq!(foo, Wrap(50));
}

```

Clippy result:

```
warning: manual implementation of an assign operation, #[warn(assign_op_pattern)] on by default
  --> src/lib.rs:16:9
   |
16 |         *self = *self * rhs
   |         ^^^^^^^^^^^^^^^^^^^
   |
help: replace it with
   |         *self *= rhs

```

Running test after applying the suggestion:

```
warning: function cannot return without recurring, #[warn(unconditional_recursion)] on by default
  --> src/lib.rs:15:5
   |
15 |     fn mul_assign(&mut self, rhs: i64) {
   |     ^
   |
note: recursive call site
  --> src/lib.rs:16:9
   |
16 |         *self *= rhs
   |         ^^^^^^^^^^^^
   = help: a `loop` may express intention better if this is on purpose

    Finished debug [unoptimized + debuginfo] target(s) in 0.76 secs
     Running target/debug/deps/assignop-22d33742ac4da442

running 1 test

thread 'test' has overflowed its stack
fatal runtime error: stack overflow

```

### PR DESCRIPTION
**Pull request by oli-obk**
fixes #1302

### CODE REVIEW

File: clippy_lints/src/assign_ops.rs [hunk 1]
```diff
@@ -135,6 +135,19 @@ impl LateLintPass for AssignOps {
                                         } else {
                                             return; // useless if the trait doesn't exist
                                         };
+                                        // check that we are not inside an `impl AssignOp` of this exact operation
+                                        let parent_fn = cx.tcx.map.get_parent(e.id);
+                                        let parent_impl = cx.tcx.map.get_parent(parent_fn);
+                                        if parent_impl.as_u32() != 0 {
```
Comment (oli-obk) [inline clippy_lints/src/assign_ops.rs]: without this check, the `get` call in the next line will panic. All other functions on `map` will also panic, so there's no `Option` way to do this correctly. Should probably fix this in rustc...

Comment (mcarton) [inline clippy_lints/src/assign_ops.rs]: That's kind of a feature, `Option<NodeId>` is not necessary as there is `syntax::ast::DUMMY_NODE_ID`. Yup, that's weak typing, but it's more memory efficient, and `NodeId`s are used everywhere.

Comment (oli-obk) [inline clippy_lints/src/assign_ops.rs]: uuuhh... once we get `panic!` inside `const fn`, we can properly implement `NonZero::new` and then we can go mad on the microoptimizations (most relevant here: `Option<NonZero<NodeId>>` being the same size as `NodeId`).

**Comment by oli-obk** [comment]
rebased with just tiny changes to address rustc changes

**Pull request merged**